File: Leetcode/NumberofLaserBeamsinaBank.rs
// https://leetcode.com/problems/number-of-laser-beams-in-a-bank

struct Solution {}

impl Solution {
    pub fn number_of_beams(bank: Vec<String>) -> i32 {
        bank.into_iter()
            .filter(|x| !x.chars().all(|c| c == '0'))
            .collect::<Vec<_>>()
            .into_iter()
            .map(|x| x.chars().filter(|x| x == &'1').count())
            .collect::<Vec<_>>()
            .windows(2)
            .fold(0, |ans, cur| ans + cur[0] * cur[1]) as i32
    }
}

File: Leetcode/NumberofLaserBeamsinaBank.cpp
// https://leetcode.com/problems/number-of-laser-beams-in-a-bank

#include <ranges>

class Solution {
   public:
    int numberOfBeams(vector<string>& bank) {
        int last = 0;
        int ans = 0;
        for (const auto& b : bank) {
            auto ok = b | std::views::filter([](char x) { return x == '1'; });
            auto l = std::ranges::distance(ok);
            if (l > 0) {
                ans += last * l;
                last = l;
            }
        }

        return ans;
    }
};